lib/libutee/base64.c

```diff
@@ -84,49 +86,46 @@ bool base64_dec(const char *data, size_t size, void *buf, size_t *blen)
 		if (!get_idx(data[n], &idx))
 			continue;
 
-		if (m >= *blen)
-			b = NULL;
-
 		switch (s) {
 		case 0:
-			if (b)
+			if (b && m < *blen)
 				b[m] = idx << 2;
 			s++;
 			break;
 		case 1:
-			if (b)
+			if (b && m < *blen)
 				b[m] |= idx >> 4;
 			m++;
-			if (m >= *blen)
-				b = NULL;
-			if (b)
-				b[m] = (idx & 0xf) << 4;
+			byte = (idx & 0xf) << 4;
 			s++;
 			break;
 		case 2:
-			if (b)
-				b[m] |= idx >> 2;
+			if (b && m < *blen)
+				b[m] = byte | (idx >> 2);
 			m++;
-			if (m >= *blen)
-				b = NULL;
-			if (b)
-				b[m] = (idx & 0x3) << 6;
+			byte = (idx & 0x3) << 6;
 			s++;
 			break;
 		case 3:
-			if (b)
-				b[m] |= idx;
+			if (b && m < *blen)
+				b[m] = byte | idx;
 			m++;
 			s = 0;
 			break;
 		default:
 			return false;	/* "Can't happen" */
 		}
 	}
-	/* We don't detect if input was bad, but that's OK with the spec. */
+
+	/*
+	 * We don't detect if input was bad, but that's OK with the spec.
+	 * We expect that each fully extracted byte is stored in output buffer.
+	 */
+	if (s == 1)
+		m++;
```

Comment:
With `s == 1` we've only extracted a partial byte in `case 0`. This doesn't seem to match the comment above.

Reply:
True.
 I think I should change the switch case above to:
```c
		(....)
		case 0:
			byte = idx << 2;
			s++;
			break;
		case 1:
			if (b && m < *blen)
				b[m] = byte | idx >> 4;
			m++;
		(...)
```
and discard this `if (s == 1) m++;`.